utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,195 @@
+import mu.KotlinLogging
+import java.time.Duration
+import java.time.Instant
+import java.util.function.Function
+
+/**
+ * This class provide ability to search messages via lw-data-provider gRPC service
+ * @param searchStep interval for sending requests via [service]. Searcher uses this value to split search interval to smaller pieces to reduce volume of requested data via one time.
+ */
+class MessageSearcher private constructor(
+    private val service: DataProviderService,
+    private val searchStep: Duration = DEFAULT_SEARCH_STEP,
+    private val defaultBook: String,
+    private val defaultSessionGroup: String,
+    private val defaultMessageStreams: Set<MessageStream>,
+) {
+    /**
+     * Searches the first message in [BASE64_FORMAT] format matched by filter. Execute search from [Instant.now] to previous time
+     * @param sessionGroup is parameter for request. Searcher uses [messageStreams] value when [sessionGroup] is null or blank
+     * @param searchInterval for searching. Searcher stops to send requests when the next `from` time is before than search start time minus [searchInterval]
+     */
+    @JvmOverloads
+    fun findLastOrNull(
+        book: String = defaultBook,
+        sessionGroup: String = defaultSessionGroup,
+        messageStreams: Set<MessageStream> = defaultMessageStreams,
+        searchInterval: Duration,
+        filter: Function<MessageSearchResponse, Boolean>
+    ): MessageSearchResponse? {
+        require(book.isNotBlank()) {
+            "'book' can't be blank"
+        }
+        require(sessionGroup.isNotBlank()) {
+            "'sessionGroup' can't be blank"
+        }
+        require(messageStreams.isNotEmpty()) {
+            "'messageStreams' can't be empty"
+        }
+        messageStreams.forEach {
+            require(it.name.isNotBlank()) {
+                "'name' in ${it.toJson()} message stream can't be blank"
+            }
+            require(it.direction != UNRECOGNIZED) {
+                "'direction' in ${it.toJson()} message stream can't be $UNRECOGNIZED"
+            }
+        }
+
+        require(!searchInterval.isZero) {
+            "'searchInterval' can't be zero"
+        }
+        require(!searchInterval.isNegative) {
+            "'searchInterval' can't be negative"
+        }
+
+        val now = Instant.now()
+        val end = now.minus(searchInterval)
+        var from: Instant
+        var to = now
+
+        val requestBuilder = createRequestBuilder(
+            book = book,
+            sessionGroup = sessionGroup,
+            messageStreams = messageStreams
+        )
+        K_LOGGER.info { "Search the last or null, base request: ${requestBuilder.toJson()}, start time: $now, end time: $end" }
+
+        return withCancellation {
+            generateSequence {
+                if (to.isBefore(end)) {
+                    return@generateSequence null
+                }
+
+                from = to
+                to = from.minus(searchStep)
```

Comment:
We can go outside the search window if the step is greater than the search interval. We should check this here and probably use `end` value in this case as a boundary